File: Csharp/MPP/protobuff/ProtoUtils.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Proto;
using proto = Proto;
using model2;
using Google.Protobuf.WellKnownTypes;

namespace protobuf
{
    static class ProtoUtils
    {
        /*
        public static BasketRequest createLoginRequest(model2.Organiser organiser)
        {
            proto.Organiser userDTO = new proto.Organiser { Id = organiser.Id, Name = organiser.Name, Passwd = organiser.Password };
            BasketRequest request = new BasketRequest { Type = BasketRequest.Types.Type.Login, User = userDTO };

            return request;
        }

        public static BasketRequest createLogoutRequest(model2.Organiser organiser)
        {
            proto.Organiser userDTO = new proto.Organiser { Id = organiser.Id, Name = organiser.Name, Passwd = organiser.Password };
            BasketRequest request = new BasketRequest { Type = BasketRequest.Types.Type.Logout, User = userDTO };

            return request;
        }

        public static BasketRequest createGetMatchesRequest()
        {
            BasketRequest request = new BasketRequest { Type = BasketRequest.Types.Type.GetMatches };
            return request;
        }

        public static BasketRequest createUpdateMatchesRequest(model2.Ticket ticket)
        {

            proto.Match matchDto = new proto.Match { Id = ticket.TicketMatch.Id };
            proto.Ticket ticketDto = new proto.Ticket { Match = matchDto, Quantity = ticket.Quantity, CustomerName = ticket.Name };

            BasketRequest request = new BasketRequest { Type = BasketRequest.Types.Type.UpdateMatches, Ticket = ticketDto };
            return request;
        }
        */

        public static BasketResponse createOkResponse()
        {
            BasketResponse response = new BasketResponse { Type = BasketResponse.Types.Type.Ok };
            return response;
        }


        public static BasketResponse createErrorResponse(String text)
        {
            BasketResponse response = new BasketResponse
            {
                Type = BasketResponse.Types.Type.Error,
                Error = text
            };
            return response;
        }

        public static BasketResponse createLoggedInResponse(model2.Organiser organiser)
        {
            proto.Organiser userDTO = new proto.Organiser { Id = organiser.Id, Name = organiser.Name, Passwd = organiser.Password };
            BasketResponse response = new BasketResponse { Type = BasketResponse.Types.Type.OrgLoggedIn, User = userDTO };

            return response;
        }

        public static BasketResponse createLoggedOutResponse(model2.Organiser organiser)
        {
            proto.Organiser userDTO = new proto.Organiser { Id = organiser.Id, Name = organiser.Name, Passwd = organiser.Password };
            BasketResponse response = new BasketResponse { Type = BasketResponse.Types.Type.OrgLoggedOut, User = userDTO };

            return response;
        }

        public static BasketResponse createGetMatchesResponse(model2.Match[] matches)
        {
            BasketResponse response = new BasketResponse
            {
                Type = BasketResponse.Types.Type.GotMatches
            };

            foreach (model2.Match m in matches)
            {
                DateTime md = DateTime.SpecifyKind(m.Date, DateTimeKind.Utc);
                var tm = Timestamp.FromDateTime(md);
                Proto.Match matchDto = new Proto.Match { Id = m.Id, Team1 = m.Team1, Team2 = m.Team2, Type = m.MatchType, NrOfSeats = m.NrOfSeats, Price = m.Price, Date = tm };


                response.Matches.Add(matchDto);
            }

            return response;
        }

        public static BasketResponse createUpdateMatchesResponse(model2.Match[] matches)
        {
            BasketResponse response = new BasketResponse
            {
                Type = BasketResponse.Types.Type.NewMatchList
            };

            foreach (model2.Match m in matches)
            {
                DateTime md = DateTime.SpecifyKind(m.Date, DateTimeKind.Utc);
                var tm = Timestamp.FromDateTime(md);
                proto.Match matchDto = new proto.Match { Id = m.Id, Team1 = m.Team1, Team2 = m.Team2, Type = m.MatchType, NrOfSeats = m.NrOfSeats, Price = m.Price, Date = tm };


                response.Matches.Add(matchDto);
            }
            return response;
        }

        public static String getError(BasketResponse response)
        {
            String errorMessage = response.Error;
            return errorMessage;
        }

        public static model2.Organiser getUser(BasketRequest request)
        {
            model2.Organiser user = new model2.Organiser(request.User.Id, request.User.Name, request.User.Passwd);
            return user;
        }

        public static model2.Organiser getUser(BasketResponse response)
        {
            model2.Organiser user = new model2.Organiser(response.User.Id, response.User.Name, response.User.Passwd);
            return user;
        }

        public static model2.Match[] getMatches(BasketResponse response)
        {
            model2.Match[] matches = new model2.Match[response.Matches.Count];
            for (int i = 0; i < response.Matches.Count; i++)
            {
                proto.Match m = response.Matches[i];
                DateTime tm = m.Date.ToDateTime();
                model2.Match match = new model2.Match { Id = m.Id, Team1 = m.Team1, Team2 = m.Team2, MatchType = m.Type, NrOfSeats = m.NrOfSeats, Price = m.Price, Date = tm };

                matches[i] = match;
            }
            return matches;
        }

        public static model2.Ticket getTicket(BasketRequest request)
        {
            model2.Match match = new model2.Match { Id = request.Ticket.Match.Id };

            model2.Ticket ticket = new model2.Ticket(request.Ticket.Id, request.Ticket.Quantity, match, request.Ticket.CustomerName);
            return ticket;
        }
    }
}

File: JavaSockets/BaskNetworking/src/main/java/bask/network/protobuffprotocol/ProtoUtils.java
package bask.network.protobuffprotocol;

import bask.model.Match;
import bask.model.Organiser;
import bask.model.Ticket;

import java.sql.Timestamp;
import java.time.Instant;

public class ProtoUtils {

    public static BasketProtobufs.BasketRequest createLoginRequest(Organiser user){
        BasketProtobufs.Organiser userDTO=BasketProtobufs.Organiser.newBuilder().setId(user.getId()).setName(user.getName()).setPasswd(user.getPassword()).build();
        BasketProtobufs.BasketRequest request= BasketProtobufs.BasketRequest.newBuilder().setType(BasketProtobufs.BasketRequest.Type.Login)
                .setUser(userDTO).build();
        return request;
    }

    public static BasketProtobufs.BasketRequest createLogoutRequest(Organiser user){
        BasketProtobufs.Organiser userDTO=BasketProtobufs.Organiser.newBuilder().setId(user.getId()).setName(user.getName()).setPasswd(user.getPassword()).build();
        BasketProtobufs.BasketRequest request= BasketProtobufs.BasketRequest.newBuilder().setType(BasketProtobufs.BasketRequest.Type.Logout)
                .setUser(userDTO).build();
        return request;
    }

    public static BasketProtobufs.BasketRequest createGetMatchesRequest()
    {
        BasketProtobufs.BasketRequest request= BasketProtobufs.BasketRequest.newBuilder()
                .setType(BasketProtobufs.BasketRequest.Type.GetMatches)
                .build();
        return request;
    }

    public static BasketProtobufs.BasketRequest createUpdateMatchesRequest(Ticket ticket)
    {
        Match match = ticket.getMatch();
//        Instant instant = match.getDate().toInstant();
//        com.google.protobuf.Timestamp date = com.google.protobuf.Timestamp.newBuilder()
//                .setSeconds(instant.getEpochSecond())
//                .setNanos(instant.getNano()).build();

        BasketProtobufs.Match matchDTO = BasketProtobufs.Match.newBuilder().setId(match.getId()).build();

        BasketProtobufs.Ticket ticketDTO=BasketProtobufs.Ticket.newBuilder().setMatch(matchDTO).setQuantity(ticket.getQuantity()).setCustomerName(ticket.getCustomerName()).build();
        BasketProtobufs.BasketRequest request= BasketProtobufs.BasketRequest.newBuilder()
                .setType(BasketProtobufs.BasketRequest.Type.UpdateMatches)
                .setTicket(ticketDTO)
                .build();
        return request;
    }

    public static BasketProtobufs.BasketResponse createOkResponse(){
        BasketProtobufs.BasketResponse response=BasketProtobufs.BasketResponse.newBuilder()
                .setType(BasketProtobufs.BasketResponse.Type.Ok).build();
        return response;
    }

    public static BasketProtobufs.BasketResponse createErrorResponse(String text){
        BasketProtobufs.BasketResponse response=BasketProtobufs.BasketResponse.newBuilder()
                .setType(BasketProtobufs.BasketResponse.Type.Error)
                .setError(text).build();
        return response;
    }

    public static BasketProtobufs.BasketResponse createLoggedInResponse(Organiser user){
        BasketProtobufs.Organiser userDTO=BasketProtobufs.Organiser.newBuilder().setId(user.getId()).setName(user.getName()).setPasswd(user.getPassword()).build();

        BasketProtobufs.BasketResponse response=BasketProtobufs.BasketResponse.newBuilder()
                .setType(BasketProtobufs.BasketResponse.Type.OrgLoggedIn)
                .setUser(userDTO).build();
        return response;
    }

    public static BasketProtobufs.BasketResponse createLoggedOutResponse(Organiser user){
        BasketProtobufs.Organiser userDTO=BasketProtobufs.Organiser.newBuilder().setId(user.getId()).setName(user.getName()).setPasswd(user.getPassword()).build();

        BasketProtobufs.BasketResponse response=BasketProtobufs.BasketResponse.newBuilder()
                .setType(BasketProtobufs.BasketResponse.Type.OrgLoggedOut)
                .setUser(userDTO).build();
        return response;
    }

    public static BasketProtobufs.BasketResponse createGetMatchesResponse(Match[] matches){
        BasketProtobufs.BasketResponse.Builder response=BasketProtobufs.BasketResponse.newBuilder()
                .setType(BasketProtobufs.BasketResponse.Type.GotMatches);
        for(Match match : matches)
        {
            Instant instant = match.getDate().toInstant();
            com.google.protobuf.Timestamp date = com.google.protobuf.Timestamp.newBuilder()
                    .setSeconds(instant.getEpochSecond())
                    .setNanos(instant.getNano()).build();

            BasketProtobufs.Match matchDTO = BasketProtobufs.Match.newBuilder().setId(match.getId()).setTeam1(match.getTeam1())
                    .setTeam2(match.getTeam2()).setType(match.getType()).setPrice(match.getPrice()).setNrOfSeats(match.getNrOfSeats()).setDate(date).build();
            response.addMatches(matchDTO);
        }
        return response.build();
    }

    public static BasketProtobufs.BasketResponse createUpdateMatchesResponse(Match[] matches)
    {
        BasketProtobufs.BasketResponse.Builder response=BasketProtobufs.BasketResponse.newBuilder()
                .setType(BasketProtobufs.BasketResponse.Type.NewMatchList);
        for(Match match : matches)
        {
            Instant instant = match.getDate().toInstant();
            com.google.protobuf.Timestamp date = com.google.protobuf.Timestamp.newBuilder()
                    .setSeconds(instant.getEpochSecond())
                    .setNanos(instant.getNano()).build();
            BasketProtobufs.Match matchDTO = BasketProtobufs.Match.newBuilder().setId(match.getId()).setTeam1(match.getTeam1())
                    .setTeam2(match.getTeam2()).setType(match.getType()).setPrice(match.getPrice()).setNrOfSeats(match.getNrOfSeats()).setDate(date).build();
            response.addMatches(matchDTO);
        }
        return response.build();
    }

    public static String getError(BasketProtobufs.BasketResponse response){
        String errorMessage=response.getError();
        return errorMessage;
    }

    public static Organiser getUser(BasketProtobufs.BasketRequest request){
        Organiser user=new Organiser();
        user.setId(request.getUser().getId());
        user.setName(request.getUser().getName());
        user.setPassword(request.getUser().getPasswd());
        return user;
    }

    public static Organiser getUser(BasketProtobufs.BasketResponse response){
        Organiser user=new Organiser();
        user.setId(response.getUser().getId());
        user.setName(response.getUser().getName());
        user.setPassword(response.getUser().getPasswd());
        return user;
    }

    public static Match[] getMatches(BasketProtobufs.BasketResponse response){
        Match[] matches=new Match[response.getMatchesCount()];
        for(int i=0;i<response.getMatchesCount();i++){
           BasketProtobufs.Match matchDTO = response.getMatches(i);

            Instant instant =  Instant.ofEpochSecond(matchDTO.getDate().getSeconds(), matchDTO.getDate().getNanos());
            Timestamp tm = Timestamp.from(instant);
            Match match = new Match(matchDTO.getId(),matchDTO.getTeam1(),matchDTO.getTeam2(),matchDTO.getType(),matchDTO.getNrOfSeats(),matchDTO.getPrice(),tm);
            System.out.println(match);
            matches[i] = match;
        }
        return matches;
    }

    public static Ticket getTicket(BasketProtobufs.BasketRequest request){
        Ticket ticket=new Ticket();
        ticket.setId(request.getTicket().getId());

        BasketProtobufs.Match matchPb = request.getTicket().getMatch();

        Instant instant =  Instant.ofEpochSecond(matchPb.getDate().getSeconds(), matchPb.getDate().getNanos());
        Timestamp tm = Timestamp.from(instant);

        Match match = new Match(matchPb.getId(),matchPb.getTeam1(),matchPb.getTeam2(),matchPb.getType(),matchPb.getNrOfSeats(),matchPb.getPrice(),tm);
        ticket.setMatch(match);
        ticket.setQuantity(request.getTicket().getQuantity());
        ticket.setCustomerName(request.getTicket().getCustomerName());
        return ticket;
    }
}
